File: synth_ai/core/tunnels/connector.py
"""cloudflared connector management with warm reuse.

This module manages the cloudflared process lifecycle:
1. Starting the connector with a tunnel token
2. Monitoring connection state
3. Keeping the connector warm between leases
4. Automatic cleanup on idle timeout

The connector is designed to be long-lived to minimize latency
for subsequent lease activations.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import re
import subprocess
import threading
import time
from datetime import datetime
from typing import Any, Optional

from .cleanup import track_process
from .errors import (
    ConnectorConnectionError,
    ConnectorNotInstalledError,
    ConnectorTokenError,
)
from .types import ConnectorState, ConnectorStatus

logger = logging.getLogger(__name__)

# Connection timeout for cloudflared
DEFAULT_CONNECTION_TIMEOUT = 60.0

# Idle timeout before stopping the connector (when no active leases)
DEFAULT_IDLE_TIMEOUT = 120.0

# Patterns to detect connection state
CONNECTED_PATTERNS = [
    re.compile(r"Registered tunnel connection", re.IGNORECASE),
    re.compile(r"Connection .* registered", re.IGNORECASE),
    re.compile(r"INF\s+.*registered", re.IGNORECASE),
]

ERROR_PATTERNS = [
    (re.compile(r"failed to connect", re.IGNORECASE), "Connection failed"),
    (re.compile(r"invalid.*token", re.IGNORECASE), "Invalid token"),
    (re.compile(r"tunnel.*not.*found", re.IGNORECASE), "Tunnel not found"),
    (re.compile(r"unauthorized", re.IGNORECASE), "Unauthorized"),
    (re.compile(r"rate.*limit", re.IGNORECASE), "Rate limited"),
]


class TunnelConnector:
    """Manages a cloudflared connector process.

    The connector is designed to be warm-reusable:
    - It stays running between lease activations
    - It automatically stops after an idle timeout
    - It can be restarted with a new token if needed

    Example:
        connector = TunnelConnector()
        await connector.start(tunnel_token)
        # ... use the tunnel ...
        await connector.stop()  # Or let it idle-timeout
    """

    def __init__(
        self,
        cloudflared_path: Optional[str] = None,
        idle_timeout: float = DEFAULT_IDLE_TIMEOUT,
    ):
        """Initialize the connector.

        Args:
            cloudflared_path: Path to cloudflared binary (auto-detected if None)
            idle_timeout: Seconds to wait before stopping on idle
        """
        self._cloudflared_path = cloudflared_path
        self._idle_timeout = idle_timeout

        self._process: Optional[subprocess.Popen[bytes]] = None
        self._state = ConnectorState.STOPPED
        self._error: Optional[str] = None
        self._connected_at: Optional[datetime] = None
        self._current_token: Optional[str] = None
        self._active_leases: set[str] = set()
        self._idle_timer: Optional[asyncio.Task[Any]] = None

        self._lock = threading.Lock()
        self._output_lines: list[str] = []
        self._output_reader_task: Optional[asyncio.Task[Any]] = None

    @property
    def status(self) -> ConnectorStatus:
        """Get the current connector status."""
        return ConnectorStatus(
            state=self._state,
            pid=self._process.pid if self._process else None,
            connected_at=self._connected_at,
            error=self._error,
        )

    @property
    def is_connected(self) -> bool:
        """Check if the connector is connected."""
        return self._state == ConnectorState.CONNECTED

    @property
    def is_running(self) -> bool:
        """Check if the process is running (may not be connected yet)."""
        return self._state in (ConnectorState.STARTING, ConnectorState.CONNECTED)

    def _get_cloudflared_path(self) -> str:
        """Get the path to cloudflared binary."""
        if self._cloudflared_path:
            return self._cloudflared_path

        # Try to find cloudflared
        from .cloudflare import get_cloudflared_path

        path = get_cloudflared_path()
        if not path:
            raise ConnectorNotInstalledError()
        return path

    async def start(
        self,
        tunnel_token: str,
        *,
        timeout: float = DEFAULT_CONNECTION_TIMEOUT,
        force_restart: bool = False,
    ) -> None:
        """Start the connector with the given tunnel token.

        If the connector is already running with the same token,
        this is a no-op (unless force_restart=True).

        Args:
            tunnel_token: The tunnel token from lease creation
            timeout: Connection timeout in seconds
            force_restart: If True, restart even if already running

        Raises:
            ConnectorNotInstalledError: If cloudflared is not installed
            ConnectorTokenError: If the token is invalid
            ConnectorConnectionError: If connection fails
        """
        # Check if already running with same token
        if self.is_running and self._current_token == tunnel_token and not force_restart:
            logger.debug("[CONNECTOR] Already running with same token")
            self._cancel_idle_timer()
            return

        # Stop existing process if running
        if self.is_running:
            await self.stop()

        cloudflared_path = self._get_cloudflared_path()
        self._state = ConnectorState.STARTING
        self._error = None
        self._connected_at = None
        self._current_token = tunnel_token
        self._output_lines.clear()

        logger.info("[CONNECTOR] Starting cloudflared connector")

        try:
            # Start cloudflared process
            cmd = [
                cloudflared_path,
                "tunnel",
                "run",
                "--token",
                tunnel_token,
            ]

            self._process = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                bufsize=1,
            )
            track_process(self._process)

            # Start output reader
            self._output_reader_task = asyncio.create_task(self._read_output())

            # Wait for connection
            await self._wait_for_connection(timeout)

            self._state = ConnectorState.CONNECTED
            self._connected_at = datetime.now()
            logger.info(
                "[CONNECTOR] Connected to Cloudflare edge (pid=%d)",
                self._process.pid,
            )

        except Exception as e:
            self._state = ConnectorState.ERROR
            self._error = str(e)
            await self._cleanup_process()
            raise

    async def _read_output(self) -> None:
        """Read output from cloudflared process."""
        if not self._process:
            return

        async def read_stream(stream: Any, name: str) -> None:
            while True:
                try:
                    line = await asyncio.get_event_loop().run_in_executor(None, stream.readline)
                    if not line:
                        break
                    decoded = line.decode("utf-8", errors="replace").strip()
                    if decoded:
                        self._output_lines.append(decoded)
                        # Keep only last 100 lines
                        if len(self._output_lines) > 100:
                            self._output_lines = self._output_lines[-100:]
                        logger.debug("[CONNECTOR] %s: %s", name, decoded)
                except Exception:
                    break

        if self._process.stdout:
            asyncio.create_task(read_stream(self._process.stdout, "stdout"))
        if self._process.stderr:
            asyncio.create_task(read_stream(self._process.stderr, "stderr"))

    async def _wait_for_connection(self, timeout: float) -> None:
        """Wait for cloudflared to connect to the edge."""
        start_time = time.time()
        poll_interval = 0.1

        while True:
            elapsed = time.time() - start_time
            if elapsed >= timeout:
                raise ConnectorConnectionError(
                    "Timeout waiting for cloudflared to connect",
                    timeout=timeout,
                )

            # Check if process died
            if self._process and self._process.poll() is not None:
                error = self._detect_error()
                if "token" in error.lower():
                    raise ConnectorTokenError(error)
                raise ConnectorConnectionError(f"cloudflared exited: {error}")

            # Check for connection in output
            for line in self._output_lines:
                for pattern in CONNECTED_PATTERNS:
                    if pattern.search(line):
                        return

                # Check for errors
                for error_pattern, error_msg in ERROR_PATTERNS:
                    if error_pattern.search(line):
                        if "token" in error_msg.lower():
                            raise ConnectorTokenError(error_msg)
                        raise ConnectorConnectionError(error_msg)

            await asyncio.sleep(poll_interval)

    def _detect_error(self) -> str:
        """Detect error from output lines."""
        for line in reversed(self._output_lines):
            for error_pattern, error_msg in ERROR_PATTERNS:
                if error_pattern.search(line):
                    return f"{error_msg}: {line}"
        return "Unknown error"

    async def stop(self) -> None:
        """Stop the connector."""
        self._cancel_idle_timer()
        await self._cleanup_process()
        self._state = ConnectorState.STOPPED
        self._current_token = None
        self._active_leases.clear()
        logger.info("[CONNECTOR] Stopped")

    async def _cleanup_process(self) -> None:
        """Clean up the cloudflared process."""
        if self._output_reader_task:
            self._output_reader_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._output_reader_task
            self._output_reader_task = None

        if self._process:
            try:
                self._process.terminate()
                # Wait briefly for graceful shutdown
                await asyncio.get_event_loop().run_in_executor(None, self._process.wait, 2.0)
            except Exception:
                with contextlib.suppress(Exception):
                    self._process.kill()
            self._process = None

    def register_lease(self, lease_id: str) -> None:
        """Register an active lease with the connector.

        This cancels any pending idle timeout.

        Args:
            lease_id: The lease ID to register
        """
        self._active_leases.add(lease_id)
        self._cancel_idle_timer()

    def unregister_lease(self, lease_id: str) -> None:
        """Unregister a lease from the connector.

        If no leases remain, starts the idle timeout.

        Args:
            lease_id: The lease ID to unregister
        """
        self._active_leases.discard(lease_id)
        if not self._active_leases:
            self._start_idle_timer()

    def _cancel_idle_timer(self) -> None:
        """Cancel the idle timeout timer."""
        if self._idle_timer:
            self._idle_timer.cancel()
            self._idle_timer = None

    def _start_idle_timer(self) -> None:
        """Start the idle timeout timer."""
        self._cancel_idle_timer()

        async def idle_shutdown() -> None:
            await asyncio.sleep(self._idle_timeout)
            if not self._active_leases:
                logger.info("[CONNECTOR] Idle timeout reached, stopping connector")
                await self.stop()

        self._idle_timer = asyncio.create_task(idle_shutdown())

    def get_logs(self, lines: int = 50) -> list[str]:
        """Get recent log lines from cloudflared.

        Args:
            lines: Number of lines to return

        Returns:
            List of recent log lines
        """
        return self._output_lines[-lines:]


# Global connector instance
_connector: Optional[TunnelConnector] = None
_connector_lock = threading.Lock()


def get_connector() -> TunnelConnector:
    """Get or create the global connector instance.

    Returns:
        TunnelConnector instance
    """
    global _connector
    with _connector_lock:
        if _connector is None:
            _connector = TunnelConnector()
        return _connector


async def ensure_connector_running(
    tunnel_token: str,
    *,
    timeout: float = DEFAULT_CONNECTION_TIMEOUT,
) -> TunnelConnector:
    """Ensure the connector is running and connected.

    Args:
        tunnel_token: The tunnel token
        timeout: Connection timeout

    Returns:
        Connected TunnelConnector instance
    """
    connector = get_connector()
    await connector.start(tunnel_token, timeout=timeout)
    return connector

File: synth_ai_core/src/tunnels/connector.rs
use std::collections::{HashSet, VecDeque};
use std::sync::Arc;
use std::time::{Duration, Instant};

use chrono::{DateTime, Utc};
use once_cell::sync::Lazy;
use parking_lot::Mutex;
use regex::Regex;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Child;
use tokio::task::JoinHandle;

use crate::tunnels::cloudflared::require_cloudflared;
use crate::tunnels::errors::TunnelError;
use crate::tunnels::types::{ConnectorState, ConnectorStatus};

static CONNECT_PATTERNS: Lazy<Vec<Regex>> = Lazy::new(|| {
    vec![
        Regex::new("Registered tunnel connection").unwrap(),
        Regex::new("Connection .* registered").unwrap(),
    ]
});

static ERROR_PATTERNS: Lazy<Vec<(Regex, &'static str)>> = Lazy::new(|| {
    vec![
        (
            Regex::new("failed to connect").unwrap(),
            "Connection failed",
        ),
        (Regex::new("invalid.*token").unwrap(), "Invalid token"),
        (
            Regex::new("tunnel.*not.*found").unwrap(),
            "Tunnel not found",
        ),
        (Regex::new("unauthorized").unwrap(), "Unauthorized"),
        (Regex::new("rate.*limit").unwrap(), "Rate limited"),
    ]
});

pub struct TunnelConnector {
    process: Option<Child>,
    state: ConnectorState,
    connected_at: Option<DateTime<Utc>>,
    error: Option<String>,
    current_token: Option<String>,
    logs: Arc<Mutex<VecDeque<String>>>,
    stdout_task: Option<JoinHandle<()>>,
    stderr_task: Option<JoinHandle<()>>,
    idle_timeout: Duration,
    active_leases: HashSet<String>,
    idle_task: Option<JoinHandle<()>>,
}

impl TunnelConnector {
    pub fn new(idle_timeout: Duration) -> Self {
        Self {
            process: None,
            state: ConnectorState::Stopped,
            connected_at: None,
            error: None,
            current_token: None,
            logs: Arc::new(Mutex::new(VecDeque::with_capacity(200))),
            stdout_task: None,
            stderr_task: None,
            idle_timeout,
            active_leases: HashSet::new(),
            idle_task: None,
        }
    }

    pub fn status(&self) -> ConnectorStatus {
        ConnectorStatus {
            state: self.state.clone(),
            pid: self.process.as_ref().and_then(|c| c.id()),
            connected_at: self.connected_at,
            error: self.error.clone(),
        }
    }

    pub fn is_connected(&self) -> bool {
        self.state == ConnectorState::Connected
    }

    pub async fn start(
        &mut self,
        token: &str,
        timeout: Duration,
        force_restart: bool,
    ) -> Result<(), TunnelError> {
        if self.state != ConnectorState::Stopped
            && self.current_token.as_deref() == Some(token)
            && !force_restart
        {
            self.cancel_idle_timer();
            return Ok(());
        }
        if self.state != ConnectorState::Stopped {
            self.stop().await?;
        }
        let bin = require_cloudflared().await?;
        let mut cmd = tokio::process::Command::new(bin);
        cmd.arg("tunnel").arg("run").arg("--token").arg(token);
        cmd.stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped());
        let mut child = cmd
            .spawn()
            .map_err(|e| TunnelError::connector(e.to_string()))?;

        let stdout = child.stdout.take();
        let stderr = child.stderr.take();
        let logs = self.logs.clone();
        if let Some(out) = stdout {
            let logs = logs.clone();
            self.stdout_task = Some(tokio::spawn(async move {
                let mut lines = BufReader::new(out).lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    push_log(&logs, &line);
                }
            }));
        }
        if let Some(err) = stderr {
            let logs = logs.clone();
            self.stderr_task = Some(tokio::spawn(async move {
                let mut lines = BufReader::new(err).lines();
                while let Ok(Some(line)) = lines.next_line().await {
                    push_log(&logs, &line);
                }
            }));
        }

        self.state = ConnectorState::Starting;
        self.error = None;
        self.current_token = Some(token.to_string());
        self.connected_at = None;
        self.process = Some(child);

        self.wait_for_connection(timeout).await?;
        self.state = ConnectorState::Connected;
        self.connected_at = Some(Utc::now());
        Ok(())
    }

    async fn wait_for_connection(&mut self, timeout: Duration) -> Result<(), TunnelError> {
        let deadline = Instant::now() + timeout;
        loop {
            if Instant::now() > deadline {
                return Err(TunnelError::connector(
                    "timeout waiting for cloudflared connection",
                ));
            }
            if let Some(proc) = &mut self.process {
                if let Ok(Some(status)) = proc.try_wait() {
                    return Err(TunnelError::connector(format!(
                        "cloudflared exited early: {status}"
                    )));
                }
            }
            {
                let logs = self.logs.lock();
                for line in logs.iter() {
                    if CONNECT_PATTERNS.iter().any(|p| p.is_match(line)) {
                        return Ok(());
                    }
                    for (p, msg) in ERROR_PATTERNS.iter() {
                        if p.is_match(line) {
                            return Err(TunnelError::connector(msg.to_string()));
                        }
                    }
                }
            }
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
    }

    pub async fn stop(&mut self) -> Result<(), TunnelError> {
        let mut proc_opt = self.prepare_stop();
        if let Some(proc) = &mut proc_opt {
            let _ = proc.start_kill();
            let _ = proc.wait().await;
        }
        Ok(())
    }

    pub fn register_lease(&mut self, lease_id: &str) {
        self.active_leases.insert(lease_id.to_string());
        self.cancel_idle_timer();
    }

    pub fn unregister_lease(&mut self, lease_id: &str) {
        self.active_leases.remove(lease_id);
        if self.active_leases.is_empty() {
            self.start_idle_timer();
        }
    }

    pub fn get_logs(&self, lines: usize) -> Vec<String> {
        let logs = self.logs.lock();
        logs.iter().rev().take(lines).cloned().collect()
    }

    fn cancel_idle_timer(&mut self) {
        if let Some(task) = self.idle_task.take() {
            task.abort();
        }
    }

    fn start_idle_timer(&mut self) {
        let timeout = self.idle_timeout;
        self.idle_task = Some(tokio::spawn(async move {
            tokio::time::sleep(timeout).await;
            let connector = get_connector();
            let mut proc_opt = None;
            {
                let mut guard = connector.lock();
                if guard.active_leases.is_empty() {
                    proc_opt = guard.prepare_stop();
                }
            }
            if let Some(mut proc) = proc_opt {
                let _ = proc.start_kill();
                let _ = proc.wait().await;
            }
        }));
    }

    fn prepare_stop(&mut self) -> Option<Child> {
        self.cancel_idle_timer();
        let proc = self.process.take();
        self.state = ConnectorState::Stopped;
        self.current_token = None;
        self.active_leases.clear();
        proc
    }
}

fn push_log(logs: &Arc<Mutex<VecDeque<String>>>, line: &str) {
    let mut guard = logs.lock();
    guard.push_back(line.to_string());
    if guard.len() > 200 {
        guard.pop_front();
    }
}

static CONNECTOR: Lazy<Mutex<TunnelConnector>> =
    Lazy::new(|| Mutex::new(TunnelConnector::new(Duration::from_secs(120))));

pub fn get_connector() -> &'static Mutex<TunnelConnector> {
    &CONNECTOR
}
